Composer/packages/client/src/recoilModel/dispatchers/navigation.ts

```diff
@@ -5,47 +5,55 @@
 //TODO: refactor the router to use one-way data flow
 import { useRecoilCallback, CallbackInterface } from 'recoil';
 import { PromptTab } from '@bfc/shared';
+import isEqual from 'lodash/isEqual';
 
 import { currentProjectIdState } from '../atoms';
 import { encodeArrayPathToDesignerPath } from '../../utils/convertUtils/designerPathEncoder';
 import { dialogsSelectorFamily, rootBotProjectIdSelector } from '../selectors';
+import { DesignPageLocation } from '../types';
 
 import { getSelected } from './../../utils/dialogUtil';
 import { designPageLocationState, focusPathState } from './../atoms/botState';
 import { checkUrl, convertPathToUrl, getUrlSearch, navigateTo } from './../../utils/navigation';
 
+const setCurrentProjectId = async ({ set, snapshot }: CallbackInterface, projectId: string) => {
+  const currentProjectId = await snapshot.getPromise(currentProjectIdState);
+  if (currentProjectId !== projectId) {
+    set(currentProjectIdState, projectId);
+  }
+};
+
 export const navigationDispatcher = () => {
   const setDesignPageLocation = useRecoilCallback(
-    ({ set }: CallbackInterface) => (projectId: string, { dialogId = '', selected = '', focused = '', promptTab }) => {
-      let focusPath = dialogId + '#';
-      if (focused) {
-        focusPath = dialogId + '#.' + focused;
-      } else if (selected) {
-        focusPath = dialogId + '#.' + selected;
-      }
-      set(currentProjectIdState, projectId);
+    (callbackInterface: CallbackInterface) => async (projectId: string, location: DesignPageLocation) => {
+      const { set, snapshot } = callbackInterface;
+
+      await setCurrentProjectId(callbackInterface, projectId);
+
+      const preLocation = await snapshot.getPromise(designPageLocationState(projectId));
+
+      if (isEqual(preLocation, location)) return;
+
+      const { dialogId = '', selected = '', focused = '' } = location;
+      const focusPath = `${dialogId}#${focused ? `.${focused}` : selected ? `.${selected}` : ''}`;
       set(focusPathState(projectId), focusPath);
```

Comment:
Here we already have 'dialogId', 'selected', 'focused' from localtion.
Then the `focusPathState` seems can be delete from client & shellData.

Reply:
Thanks, @zhixzhan I will create a new PR to clean the data,